commit dcc16cfaa1db6984b255ab2768697bd675e64dcd
Author: Andy Whitcroft <apw@canonical.com>
Date:   Tue May 1 16:17:52 2012 +0100

    UBUNTU: ubuntu: overlayfs -- overlayfs: switch to use inode_only_permissions
    
    When checking permissions on an overlayfs inode we do not take into
    account either device cgroup restrictions nor security permissions.
    This allows a user to mount an overlayfs layer over a restricted device
    directory and by pass those permissions to open otherwise restricted
    files.
    
    Switch over to the newly introduced inode_only_permissions.
    
    Signed-off-by: Andy Whitcroft <apw@canonical.com>
    Signed-off-by: Miklos Szeredi <mszeredi@suse.cz>

diff --git a/fs/overlayfs/inode.c b/fs/overlayfs/inode.c
index 86bf663..1a8e232 100644
--- a/fs/overlayfs/inode.c
+++ b/fs/overlayfs/inode.c
@@ -105,19 +105,9 @@ int ovl_permission(struct inode *inode, int mask)
 		if (is_upper && !IS_RDONLY(inode) && IS_RDONLY(realinode) &&
 		    (S_ISREG(mode) || S_ISDIR(mode) || S_ISLNK(mode)))
 			goto out_dput;
-
-		/*
-		 * Nobody gets write access to an immutable file.
-		 */
-		err = -EACCES;
-		if (IS_IMMUTABLE(realinode))
-			goto out_dput;
 	}
 
-	if (realinode->i_op->permission)
-		err = realinode->i_op->permission(realinode, mask);
-	else
-		err = generic_permission(realinode, mask);
+	err = inode_only_permission(realinode, mask);
 out_dput:
 	dput(alias);
 	return err;
